Add Prepare stage for outputs

Aflynn50 opened the pull request:
This PR adds the prepare stage for outputs. The prepare function takes instansiations of all the types mentioned in the output expressions of the query. The output expressions are checked against these types. If an output expression contains a * this is substituted for a list of all the tags in the output expression struct.

All output expression columns are given an alias with prefix `_sqlair` and suffix `_n` where `n` is the position of the output in the query. This will be used in the scan/decode stage to retrieve the relevant columns.

solmyrcan commented on internal/expr/expr_test.go:
```diff
@@ -297,3 +314,144 @@ func FuzzParser(f *testing.F) {
+
+	for i, test := range testList {
+		parser := expr.NewParser()
+		parsedExpr, err := parser.Parse(test.sql)
```
Perhaps change to `parsedExpr, _ :=` , since `err` gets over-written in the next line.

TheSignPainter98 replied:
Alternatively, I think asserting that `err` is nil could also be good here :)

Aflynn50 replied:
Actually it'll probably be better if I put this in
```golang
if err != nil {
	c.Fatal(err)
}
```
Since we get a null pointer expression for some reason that did happen to be an error.

solmyrcan replied:
sounds good!

solmyrcan commented on internal/expr/expr_test.go:
```diff
@@ -297,3 +314,144 @@ func FuzzParser(f *testing.F) {
+		parsedExpr, err := parser.Parse(test.sql)
+		_, err = parsedExpr.Prepare(test.structs...)
+		c.Assert(err, ErrorMatches, `cannot prepare expression: unknown type: .*`,
+			Commentf("test %d failed:\nsql: '%s'\nstructs: '%+v'\np:'%s'", i, test.sql, test.structs, parsedExpr.String()))
```
Perhaps change `p:'%s'` to something bit more readable like `parsed to: '%s'`

Aflynn50 replied:
Ah, that was only there for my debugging! I'll remove it.

solmyrcan commented on internal/expr/expr_test.go:
```diff
@@ -297,3 +314,144 @@ func FuzzParser(f *testing.F) {
+
+	for i, test := range testList {
+		parser := expr.NewParser()
+		parsedExpr, err := parser.Parse(test.sql)
```
Same here - perhaps change to `parsedExpr, _ :=` , since `err` gets over-written in the next line. If it makes sense, then it's worth checking the rest of the file for similar lines :)

Aflynn50 replied:
Same comment as above.

solmyrcan commented on internal/expr/expr_test.go:
```diff
@@ -297,3 +314,144 @@ func FuzzParser(f *testing.F) {
+		sql     string
+		structs []any
+	}{{
+		sql:     "SELECT street FROM t WHERE x = $Address.street",
```
Would it make sense to add a one-liner brief comment for each of these test cases to say why they must fail?

Aflynn50 replied:
Hopefully the error message should have enough context for these. I've just split this one into two separate tests so that the test titles give a bit more context too.

solmyrcan commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
 }
 
+// parseColumn parses a column made up of name bytes, optionally dot-prefixed by
+// its table name. parseColumn returns an error so that it can be used with
```
The code doesn't seem to return an error :) Perhaps we need to replace `nil` with an `error` in error cases.

Aflynn50 replied:
Ah this is saying it returns an error type, i.e. the parseList function takes this function as an argument so the type that parseList wants must match the type of this function. Even though we never actually need to return an error from this function.

solmyrcan commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+	cp := p.save()
+
+	if !p.skipByte('(') {
+		return nil, false, nil
```
Should we return an error here as done elsewhere, replacing the last `nil` with an `fmt.Errorf()` with an relevant error message?

Aflynn50 replied:
When we return false but not an error in the parser its because we've not found the thing we are looking for. If we return an error then we have found it but we thing it is incorrectly formatted.
Here we return `nil, false` if we cant find an opening bracket that signifies the start of the list, i.e. we cant find a list to parse.

solmyrcan commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+		} else if i == 0 {
+			// If the first item is not what we are looking for, we exit.
+			cp.restore()
+			return nil, false, nil
```
Should we return an error here as done elsewhere, replacing the last `nil` with an `fmt.Errorf()` with an relevant error message?

Aflynn50 replied:
See comment above.

solmyrcan commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+	}
+
+	cp.restore()
+	return nil, false
```
Would this method need to return an error when both `parseColumn()` and `parseList()` do not return a happy result? i.e. when both Case 1 and Case 2 are not applicable. Or, similar to what's done at `parseTargets()` below, would it need to return within Case 1 and Case 2 when an error occurs?

Aflynn50 replied:
See comment above.

solmyrcan commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+}
+
+// parseTargets parses the part of the output expression following the
+// ampersand. This can be one or more references Go objects.
```
Change to "Go object references" maybe? :)

Aflynn50 replied:
Changed.

solmyrcan commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+
+// parseTargets parses the part of the output expression following the
+// ampersand. This can be one or more references Go objects.
+// If the ampersand is not preceded by a space and succeeded by a name or
```
Should this be "preceded by a space or an opening bracket and succeeded by a name" instead?

Aflynn50 replied:
True, the syntax has changed since this was written.

solmyrcan commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+
+	// Case 1: A single target e.g. "&Person.name".
+	if target, ok, err := p.parseTarget(); err != nil {
+		return nil, false, err
```
Would we need to call `cp.restore()` before returning? Same thing in Case 2.

Aflynn50 replied:
No, because the state hasnt actually changed if we fail. Also its an error case here so the state is discarded anyway.

solmyrcan commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+
+	// Case 1: There are no columns e.g. "&Person.*".
+	if targets, ok, err := p.parseTargets(); ok {
+		return &outputPart{[]fullName{}, targets}, true, nil
```
Do we need to call `cp.restore()` before each return statement in this method?

Aflynn50 replied:
Its not actually needed here, since the functions dont change the state.

solmyrcan commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,183 @@
+// prepareInput checks that the input expression corresponds to a Go struct
+// passed to Prepare.
+func prepareInput(ti typeNameToInfo, p *inputPart) error {
+	if inf, ok := ti[p.source.prefix]; ok {
```
Could this method be modified to follow the style of first checking the errors and then leaving the success case as the last one :)

Aflynn50 replied:
Indeed it can, and it has! I just forgot to merge the branch it was done in.

solmyrcan commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,183 @@
+	// Check asterisk are in correct places.
+
+	sct := starCount(p.target)
+	scc := starCount(p.source)
```
`scs` (`s` for source) instead? :P

Aflynn50 replied:
That does make sense. I was debating between the two and chose c for columns but I think s might be better.

solmyrcan commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	starTarget := sct == 1
+	starColumn := scc == 1
```
Would `starSource` be a better fit as a name here?

Aflynn50 replied:
Yea probably. I've changed it.

solmyrcan commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,183 @@
+			pref := ""
+
+			// Prepend table name. E.g. "t" in "t.* AS &P.*".
+			if lenS > 0 && p.source[0].prefix != "" {
```
Do we need to return any error if `lenS >0` but there's no prefix? Or has it been checked at this point?

Aflynn50 replied:
Ah yes, I think this should actually be `if starSource`

TheSignPainter98 commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
 	return "", false
 }
 
+// parseColumn parses a column made up of name bytes, optionally dot-prefixed by
```
I think 'column made up of name bytes' could be shortened to 'a column name'

Aflynn50 replied:
Name bytes here refers to the specific list of bytes that can make up names (which right now is particular to this library)

TheSignPainter98 commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+func (p *Parser) parseTarget() (fullName, bool, error) {
+	cp := p.save()
+
+	if p.skipByte('&') {
```
I might be missing context here, but I recall that target names should be proceded by ` &`, does this need to handle the space? Or should it be handled elsewhere?

Aflynn50 replied:
It's being handled in the advance function, which has yet to be approved/reviewed! I'll change this comment in that PR.

TheSignPainter98 commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+
 // parseGoFullName parses a Go type name qualified by a tag name (or asterisk)
-// of the form "TypeName.col_name". On success it returns the parsed fullName,
+// of the form "&TypeName.col_name". On success it returns the parsed FullName,
```
If my comment above about ` &` is true, does this need to to mention the preceding space?

Aflynn50 replied:
See above.

TheSignPainter98 commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
-// parseInputExpression parses an input expression of the form $Type.name.
+// parseList takes a parsing function that returns a fullName and parses a
+// bracketed, comma seperated, list.
+func (p *Parser) parseList(parseFn func(p *Parser) (fullName, bool, error)) ([]fullName, bool, error) {
```
This might be over-engineering somewhat, but I wonder whether the need to parse a comma-separated (or indeed _something_-separated) list may arise elsewhere. Perhaps this function could be generalised?
```go
func (p *Parser) parseList(parseFn func(p *Parser) (any, bool, error)) ([]any, bool, error)
```
Generics could be nice, but Go doesn't seem to support generic methods (yet?)

Aflynn50 replied:
It probably would, though I doubt this library will have any further need of it. I would also hope that someone someone has already done this better haha.

TheSignPainter98 commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+		}
+
+		nextItem = p.skipByte(',')
+		p.skipSpaces()
```
Just to save one line of code, could this `p.SkipSpaces()` be moved to the top of the loop, then the `p.SkipSpaces()` line be removed?

Aflynn50 replied:
True, it does change the behaviour slightly but since it enters an error state when it exits the loops I think its ok.

TheSignPainter98 commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+}
+
+// parseOutputExpression parses all output expressions. The ampersand must be
+// preceded by a space and followed by a name byte.
```
I think 'byte' points to an implementation detail, maybe just 'followed by a name' would be clearer?

Also, this function doesn't list its failure conditions, like others :P

Aflynn50 replied:
See above

TheSignPainter98 commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+// parseTargets parses the part of the output expression following the
+// ampersand. This can be one or more references Go objects.
+// If the ampersand is not preceded by a space and succeeded by a name or
+// opening bracket the targets are not parsed.
```
I'm not sure this note about the spaces and ampersand belongs here? This function doesn't parse these spaces itself, so perhaps it belongs elsewhere?

Also, I don't think this forces that a space should come before the first target? Is `SELECT&Person.name` rejected elsewhere?

Aflynn50 replied:
Removed.

TheSignPainter98 commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+	return nil, false, nil
+}
+
+// parseOutputExpression parses all output expressions. The ampersand must be
```
I think 'all' here is somewhat misleading, to me, it implies that this function could also parse expressions later on, even when they are out of place. Perhaps something like 
```go
// parseOutputExpression parses an output expression which may consist of multiple parts. ...
```

Also, I think "The ampersand _before identifiers_ must be preceded..." could be clearer perhaps?

Aflynn50 replied:
Changed.

TheSignPainter98 commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+		return &outputPart{[]fullName{}, targets}, true, nil
+	} else if err != nil {
+		return nil, false, err
+	}
```
Just a minor re-ordering note, should `err` be checked first and `ok` second?

Aflynn50 replied:
Changed.

TheSignPainter98 commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+	if cols, ok := p.parseColumns(); ok {
+		p.skipSpaces()
+		if p.skipString("AS") {
+			p.skipSpaces()
```
Does the return value of the second `skipSpaces` call need to be read? Otherwise, I don't think the space-before-ampersands condition is checked

Aflynn50 replied:
True, this will be tightened up once the advance is in.

TheSignPainter98 commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,183 @@
+	// Check asterisk are in correct places.
+
+	sct := starCount(p.target)
+	scc := starCount(p.source)
```
I'm having a hard time reading the code below, perhaps these could be renamed `targetStars` and `sourceStars`?

Aflynn50 replied:
That might get confusing with `starTarget` and `starSource`. Since their scope is only a few lines I think its ok as it is.

TheSignPainter98 commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	starTarget := sct == 1
+	starColumn := scc == 1
```
I think there's a small mismatch with naming with `starColumn`, if `scc` is inlined, the variable and field name differ:
```go 
starColumn := starCount(p.source)
```

Aflynn50 replied:
Changed.

TheSignPainter98 commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,183 @@
+	starColumn := scc == 1
+
+	lenS := len(p.source)
+	lenT := len(p.target)
```
I think `numSources` and `numTargets` might be clearer?

Aflynn50 replied:
Changed.

TheSignPainter98 commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,183 @@
+				pref = p.source[0].prefix + "."
+			}
+
+			for tag := range inf.tagToField {
```
Perhaps it's because I haven't seen this form so much, but instead of implicitly ignoring the value, is it clearer to write `for tag, _ := range ...`? I'm honestly not sure on this one, short is nice too

Aflynn50 replied:
I think the var names make it clear enough whats going on here. The unclear thing here is Go's for loop syntax tbh.

TheSignPainter98 commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,183 @@
+func (pe *ParsedExpr) Prepare(args ...any) (expr *PreparedExpr, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot prepare expression: %s", err)
```
I think I recall seeing this in an older PR, it still seems a little magical to me (and hence un-Go-like?)

Would it be cleaner to have a function to do this (name tbd) and to call it explicitly each time?
```go
func (pe *ParsedExpr) Prepare(...) (...) {
    errFmt := func(e error) error {
        return fmt.Errorf("cannot prepare expression: %s", e)
    }
    ...
        return nil, errFmt(err)
}
```

Aflynn50 replied:
It will, since both parse and prepare are actually in the internal subdirectory we will have a nice user friendly wrapper round the outside that brushes over confusing names like parse prepare and complete. But thats all for the future.

TheSignPainter98 commented on internal/expr/typeinfo_test.go:
```diff
@@ -3,14 +3,10 @@ package expr
 )
 
-// Hook up gocheck into the "go test" runner.
-func TestInternal(t *testing.T) { TestingT(t) }
```
I assume this has been replaced elsewhere ;)

Aflynn50 replied:
Yes, it turns out you only need one for both the internal and external testing, even though they're in different packages (`expr` vs `expr_test`). All the test results were appearing twice.

niemeyer commented on internal/expr/parts.go:
```diff
@@ -10,7 +14,7 @@ type queryPart interface {
 
 	// ToSQL returns the SQL representation of the part.
-	toSQL() string
+	toSQL([]string, int) string
```
As we discussed during the call, you're right that this is a bad smell. Let's just kill this method and move that trivial logic to the call site, and implement a private `part()` marker method that does the validation work.

Aflynn50 replied:
I changed this a few days ago, forgot to reply here!

Aflynn50 committed:
Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Change error wording

Tidy up comments

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Change error wording

Tidy up comments

Var name change

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Change error wording

Tidy up comments

Var name change

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Change error wording

Tidy up comments

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Change error wording

Tidy up comments

Var name change

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Change error wording

Tidy up comments

Var name change

Fix tests

Change comment

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Change error wording

Tidy up comments

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Change error wording

Tidy up comments

Var name change

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Change error wording

Tidy up comments

Var name change

Fix tests

Change comment

Remove accidental output parser changes

Added id generation (#11)

* Added id generation

* Add the column name to the unique string

* Updated tests and chaged prepare tests to use regex

Remove old comment

Disallow mixed types in output expression

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Change error wording

Tidy up comments

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Change error wording

Tidy up comments

Var name change

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Add prepare output

Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

fixed up tests

Tests half fixed

Fix tests

Minor refactor and comment update

Updated output parser with go style guide advice

Add star checking

Add testing for asterisks

Restructure tests

Change comment and remove useless cp.save

Change 'near char x' to 'column x'

Optimise cp.save in parseList

Removed unneeded cp.save/cp.restore

Remove extraneous comments

Add 'to' in comment

Change comments

Refactor prepareInput

Change error wording

Tidy up comments

Var name change

Fix tests

Change comment

Remove accidental output parser changes

Added id generation (#11)

* Added id generation

* Add the column name to the unique string

* Updated tests and chaged prepare tests to use regex

Remove old comment

Disallow mixed types in output expression

Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Refactor prepareInput

Change error wording

Tidy up comments

Remove duplicate test

Fix rebase conflicts

Aflynn50 committed:
Combine lines

Aflynn50 committed:
Rebased

Aflynn50 committed:
Change outputs to be just info

Aflynn50 committed:
Adjust outputs type

Aflynn50 committed:
Update getKeys and add comments

Aflynn50 committed:
Improve error messages

Aflynn50 committed:
Refactored star check

Aflynn50 committed:
Tidy up

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -41,7 +82,91 @@ func prepareInput(ti typeNameToInfo, p *inputPart) error {
+				return nil, fmt.Errorf(`type %s unknown, have: %s`, t.prefix, strings.Join(getKeys(ti), ", "))
+			}
+		} else if t.prefix != info.structType.Name() {
+			return nil, fmt.Errorf("multiple target types in: %s", p.String())
```
As we discussed, looks like we could handle this well as it's just a different spelling of the individual parts (`one AS &Foo.one, two AS &Bar.two` vs. `(one, two) AS (&Foo.one, &Bar.two)`).

Aflynn50 replied:
Changed, this is supported now.

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -41,7 +82,91 @@ func prepareInput(ti typeNameToInfo, p *inputPart) error {
+		}
+	}
+
+	// Case 2: None star target cases e.g. "...&(P.name, P.id)".
```
Typo in the comment I think (`(&P...` vs `&(P...`). Same below.

Aflynn50 replied:
Fixed.

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -77,12 +204,30 @@ func (pe *ParsedExpr) Prepare(args ...any) (expr *PreparedExpr, err error) {
+			for i, c := range outCols {
+				sql.WriteString(c.String())
+				sql.WriteString(" AS _sqlair_")
+				sql.WriteString(c.name)
```
Per conversation, I think we can use just `_sqlair_1`, without the name, and then have a private slice that maps from those indexes into the fields.

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/expr_test.go:
```diff
@@ -36,25 +36,23 @@ type District struct {
-	`SELECT "'" AS &Person.*`,
-	`[Bypass[SELECT "'" AS ] Output[[] [Person.*]]]`,
+	[]any{Person{}},
+	"SELECT p.address_id AS [a-zA-Z_0-9]+, p.id AS [a-zA-Z_0-9]+, p.name AS [a-zA-Z_0-9]+",
```
Given the other comment about simplifying the field name to `_sqlair_1` and so on, it would be nice to see those in the tests so we can both understand how it works here and also have logic that ensures our logic is right.

Aflynn50 replied:
Changed. They now don't use regex.

niemeyer reviewed (changes requested):
Looking good! Just a few details:

Aflynn50 committed:
Added support for mixed types e.g. (&P.id, &A.di) and removed tags
from _sqlair alias

Aflynn50 committed:
Changed outputDest info in light of spike development

niemeyer commented on internal/expr/expr_test.go:
```diff
@@ -36,25 +36,23 @@ type District struct {
-	`SELECT "'" AS &Person.*`,
-	`[Bypass[SELECT "'" AS ] Output[[] [Person.*]]]`,
+	[]any{Person{}},
+	"SELECT p.address_id AS _sqlair_0, p.id AS _sqlair_1, p.name AS _sqlair_2",
```
Suddenly it starts to feel real. :)

niemeyer commented on internal/expr/expr_test.go:
```diff
@@ -326,44 +413,136 @@ func FuzzParser(f *testing.F) {
+		}
+		_, err = parsedExpr.Prepare(test.structs...)
+		c.Assert(err, ErrorMatches, `cannot prepare expression: type [a-zA-Z_\.0-9]+ unknown, have: .*`,
+			Commentf("test %d failed:\nsql: '%s'\nstructs: '%+v'", i, test.sql, test.structs))
```
It would be nice to have the real error here, as that's the only way we'll know the error message is indeed correct. As it is, even if it's right today, we can break it tomorrow and not notice.

For that, have you considered adding an `error string` field on the big testing list above, and if it's set, we fail unless an error is found, and do `ErrroMatches` with the set string?

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/expr_test.go:
```diff
@@ -326,44 +413,136 @@ func FuzzParser(f *testing.F) {
+			c.Fatal(err)
+		}
+		_, err = parsedExpr.Prepare(test.structs...)
+		c.Assert(err, ErrorMatches, `cannot prepare expression: type [a-zA-Z_\.0-9]+ has no "[a-zA-Z_\.0-9]+" db tag`,
```
Same here.

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/expr_test.go:
```diff
@@ -326,44 +413,136 @@ func FuzzParser(f *testing.F) {
 		}
-		c.Assert(preparedExpr.SQL, Equals, test.expectedPrepared)
+		_, err = parsedExpr.Prepare(test.structs...)
+		c.Assert(err, ErrorMatches, "cannot prepare expression: invalid asterisk in output expression: .*",
```
Same.

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/expr_test.go:
```diff
@@ -326,44 +413,136 @@ func FuzzParser(f *testing.F) {
+			c.Fatal(err)
+		}
+		_, err = parsedExpr.Prepare(test.structs...)
+		c.Assert(err, ErrorMatches, "cannot prepare expression: invalid asterisk in output expression: .*",
```
Same.

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/expr_test.go:
```diff
@@ -326,44 +413,136 @@ func FuzzParser(f *testing.F) {
+			c.Fatal(err)
+		}
+		_, err = parsedExpr.Prepare(test.structs...)
+		c.Assert(err, ErrorMatches, "cannot prepare expression: mismatched number of cols and targets in output expression: .*",
```
Same.

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -41,7 +79,99 @@ func prepareInput(ti typeNameToInfo, p *inputPart) error {
 }
 
-// Prepare takes a parsed expression and all the Go objects mentioned in it.
+// prepareOutput checks that the output expressions are correspond to a known types.
```
Some typos in this sentence.

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -41,7 +79,99 @@ func prepareInput(ti typeNameToInfo, p *inputPart) error {
+	var outCols = make([]fullName, 0)
+	var outDests = make([]outputDest, 0)
+
+	// Check the asterisk are well formed (if present).
```
s/the asterisk/that asterisks/

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -41,7 +79,99 @@ func prepareInput(ti typeNameToInfo, p *inputPart) error {
+			// For a star expression we fill out the destinations as we generate the columns.
+			outDests = append(outDests, outputDest{info.structType, f})
+
+		}
```
Nitpick: extra spaces before closings are generally not visually helpful.

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -41,7 +79,99 @@ func prepareInput(ti typeNameToInfo, p *inputPart) error {
+	if p.target[0].name == "*" {
+		info, _ := ti[p.target[0].prefix]
+
+		// Case 1.1: Single star i.e. "t.* AS &P.*" or "&P.*"
```
Nice comments all around here, thanks.

niemeyer reviewed (approved):
Thanks, Alastair. The comments here are pretty much all cosmetic now, so I'm going ahead and merging this PR as it's blocking the follow ups from being clean. Can you please apply these trivials on an independent PR later?  Thanks again.

Aflynn50 committed:
Made error's explicit and corrected comments
